fixes rod duplication bug with reinforced floors and disposals

AlinaStarkova opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

fixes the infinate rod bug where disposal pipes ejecting things under a reinforced floor will break off a rod, but not actually remove the reinforcement, allowing for loops to get infinate rods and by extension metal material. this adds a simple check to force reinforced floors to break when this occurs

by default any floor that overrides make plating in the future will not pop a tile here, but you can set force_plating = TRUE on the tile to do so anyway, this also will not try to pop glass floors (as the pipe is over it, not under it, also if it did it qdels the floor for some reason). everythings tested
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
fixes #27256 


https://user-images.githubusercontent.com/40489693/104872842-f1770e80-591c-11eb-9c72-f9960a307712.mp4


<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: Nari Harimoto
fix: fixed the rod duplication bug using reinforced floors and disposals. disposal ejections now properly break floor reinforcement, dropping you just 1 rod.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

AlinaStarkova committed:
fixes infinate rod bug

fixes the infinate rod bug where disposal pipes ejecting things under a reinforced floor will break off a rod, but not actually remove the reinforcement, allowing for loops to get infinate rods and by extension metal material. this adds a simple check to force reinforced floors to break when this occurs

LemonInTheDark commented on code/modules/recycling/disposal/pipe.dm:
```diff
@@ -97,7 +97,10 @@
 		if(floorturf.floor_tile)
 			new floorturf.floor_tile(T)
-		floorturf.make_plating()
+		if(istype(floorturf, /turf/open/floor/engine))
+			floorturf.make_plating(TRUE)
+		else
+			floorturf.make_plating()
```
is there any case where you wouldn't want the tile to break?

AlinaStarkova replied:
yes, space tiles and open space tiles are technically considered floors. and those 2, as well as reinforce floor, is the only floors which have make_plating be false. obviously we dont want to run make plating on space or openspace, only reinforced floor, so we dont want to force it for every floor tile

Cyberboss replied:
It feels like a situation that could apply elsewhere. Maybe consider handling it in a new `/turf/open/floor` proc?

Cyberboss replied:
Or a new arg to `make_plating`

AlinaStarkova replied:
ill try

AlinaStarkova replied:
made a proc

LemonInTheDark replied:
> 
> 
> yes, space tiles and open space tiles are technically considered floors. and those 2, as well as reinforce floor, is the only floors which have make_plating be false. obviously we dont want to run make plating on space or openspace, only reinforced floor, so we dont want to force it for every floor tile

Are you sure about that? make_plating is defined on /turf/open/floor, and space tiles are /turf/open/space.
We only care about force on reinforced floors, since they're the only thing that overrides make_plating outside of platings. Wouldn't just passing in true be acceptable here?

AlinaStarkova replied:
my bad, not space, i was tired when i said that, i meant plating, plating is considered a floor tile and we dont want that to have make_plating called for it, glass floors also dont want that called for it, as those function visually like plating as well @LemonInTheDark

AlinaStarkova replied:
also, i wanted to future proof it in case theres anything we dont want to make plating but count as a floor tile, i can think of a few examples, such as a liquidy tile where you just fly through the tile without breaking it, or a non-physical tile. idk, this way is future proofed and i believe thats going to be better in the long run

LemonInTheDark replied:
I can get behind adding some way of preventing floor_tile creation, but wouldn't that be better served by a proc? there are multiple bits of the code that do something similar to this, so a more general solution would be better.

As for make_plating, if something really just doesn't want to be made into plating, it can override the proc in the same way `/turf/open/floor/plating` does. I suppose force is rather poorly named, as it is only used by `/turf/open/floor/engine/make_plating()`, it was added to solve a similar issue with singularity pulling a couple months back.

AlinaStarkova replied:
i dont think a proc to prevent floor tile creation is the best solution here since i dont want it to be prevented on a tiles subtype inherently, only in instances in which a tile is set to not break, this can occur in multiple times and not just with the engine tile. for example: indestructible versions of tiles wont break, so every indestructible subtype needs to be able to not pop if its not forced to break

LemonInTheDark replied:
Wouldn't it be better to make a proc that spawns floor tiles, then replace everything that spawns floor tiles with the proc, then override the proc to nothing if something doesn't want to spawn floor tiles? This all ignores that things that don't want to spawn floor tiles can just set their floor tile to null and it'll work just fine.

AlinaStarkova committed:
simplified code

new proc and var to set if the plating can be forced to break or not

AlinaStarkova committed:
fixed the var used

AlinaStarkova committed:
added pop_tile

by default floor tiles can be popped by pipes, set a floor tiles var to FALSE if you dont want it popped by pipes

pipe popping overrides popping to TRUE, and code changed to only drop the stack item of the floor tile *IF* pop_tile is true..... this change was done so that floor tiles wont spawn their stack item despite not being able to be forced to plating (currently only effects reinforced floors)

AlinaStarkova commented:
in testing i found that even when var editing the reinforced floors forceed_plating var to FALSE, it still spawned the stack item rod despite not being able to pop, hence i added the pop_tile var which by default is true, but is currently false on plating and reinforced floors, if the floor tiles forced_plating var is TRUE, the pipe will automatically set pop_tile var to TRUE as well, allowing it to spawn is stack item and break the tile, else if forced_plating is FALSE then it will neither break the tile or spawn its stack item

AnturK commented:
What's going on with this ?

AlinaStarkova commented:
> What's going on with this ?

i got fired from my job the day after i made this PR so i took a break from coding for a bit. ill work on this later tonight possibly

(it technically works as is, but lemons wants me to make the tile popping a proc even though it only needs to be checked on the pipe itself.)

LemonInTheDark reviewed (changes requested):
Sorry for being so all over the place, hope this is less of a mess

AlinaStarkova commented on code/game/turfs/open/floor.dm:
```diff
@@ -44,6 +46,8 @@
 	if(is_station_level(z))
 		GLOB.station_turfs += src
 
+/turf/open/floor/proc/force_plating(forced_plating)
+	make_plating(forced_plating)
```
because other floor tiles can be made that will require it, and i dont think it good practice to override it here every single time you add a new floor that normally doesnt want to break, just because this only applies to one tile now, doesnt mean it always will, and i was specifically told by another maintainer to use a proc

AlinaStarkova commented on code/modules/recycling/disposal/pipe.dm:
```diff
@@ -95,9 +95,11 @@
-		if(floorturf.floor_tile)
+		if(floorturf.forced_plating == TRUE)
+			floorturf.pop_tile = TRUE
+		if(floorturf.floor_tile && floorturf.pop_tile == TRUE)
 			new floorturf.floor_tile(T)
```
no, because this is only occuring *in this instance* ... i dont want to override the tile directly because this will not apply in any and all instances, *only* in this instance (a disposals pipe busting through the floor) will this ever occur, so overriding it on the stuff directly will mess with normal functioning behaviour

Qustinnus commented on code/modules/recycling/disposal/pipe.dm:
```diff
@@ -95,9 +95,11 @@
+		if(floorturf.floor_tile && floorturf.pop_tile == TRUE)
 			new floorturf.floor_tile(T)
-		floorturf.make_plating()
+		floorturf.force_plating(floorturf.forced_plating)
```
```suggestion
		floorturf.make_plating(floorturf.forced_plating)
```
kill the useless proc

Qustinnus commented on code/game/turfs/open/floor.dm:
```diff
@@ -44,6 +46,8 @@
 	if(is_station_level(z))
 		GLOB.station_turfs += src
 
+/turf/open/floor/proc/force_plating(forced_plating)
+	make_plating(forced_plating)
```
```suggestion
```

AlinaStarkova committed:
changes to simpler code

it simply checks if its a floor turf and intact (aka, not walls, and not glass floors (else it will delete the glass floor)) and if so, it will make plating if the floor allows it to make plating in this instance and pops its stack item if it has one

LemonInTheDark commented on code/modules/recycling/disposal/pipe.dm:
```diff
@@ -93,11 +93,11 @@
+	if(isfloorturf(T) && T.intact) //intact floor, pop the tile
 		floorturf = T
-		if(floorturf.floor_tile)
+		if(floorturf.floor_tile && floorturf.forced_plating == TRUE)
```
```suggestion
		if(floorturf.floor_tile && floorturf.make_plating(floorturf.forced_plating))
```
make_plating returns null if it fails and true if it actually makes plating. You should take advantage of that here.

On another note, what's the purpose of forced_plating?

AlinaStarkova replied:
its a check to ensure this is only being forced when its supposed to be forced, in testing, even if i did make_plating(force) floors were still poping their stack item even though they werent breaking, the var is an added check to ensure this isnt happening if its not actually breaking its tile.

LemonInTheDark replied:
ah then checking the return value of make plating is a suitable replacement

AlinaStarkova replied:
i couldnt figure out how to check the return value, no matter how you explained it about it returning null, i couldnt see *where* it said it returned null nor how to even check for that. and in testing that method i realized checking if a tile can make plating before spawning its stack item wasnt working quite as intended (it worked, but some tiles stopped spawning their stack item while some did) so im just going to force it at all times and have this be a problem for whoever makes the first floor that cant be popped from disposals pipes

AlinaStarkova committed:
just going to force it at all times

if people make a new tile that this doesnt want to be forced they can code it, i ran into some issues about actually spawning the stack item not working properly regardless of checks about wether the floor can pop or not, so its just going to be forced at all times
